Use `cardano-ledger` for all minimum UTxO calculations

jonathanknowles opened the pull request:
## Issue Number

ADP-2144

## Summary

This PR:

-  uses the ledger function [`evaluateMinLovelaceOutput`](https://github.com/input-output-hk/cardano-ledger/blob/68a535603c80b7cf53425fd34558e23f9983dfe8/eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Wallet.hs#L506) to replace the use of Cardano API function [`calculateMinimumUTxO`](https://github.com/input-output-hk/cardano-node/blob/42809ad3d420f0695eb147d4c66b90573d27cafd/cardano-api/src/Cardano/Api/Fees.hs#L1226) in all implementation code, for all eras.
- repurposes the Cardano API function [`calculateMinimumUTxO`](https://github.com/input-output-hk/cardano-node/blob/42809ad3d420f0695eb147d4c66b90573d27cafd/cardano-api/src/Cardano/Api/Fees.hs#L1226) for use as an oracle, to compare against when testing.
- removes all special-casing for the Babbage era, and replaces it with a simpler implementation that works in all eras.
- revises documentation for the `{compute,isBelow}minimumCoinForUTxO` functions to explain their intended purpose more clearly.

## Context

- The ledger function [`evaluateMinLovelaceOutput`](https://github.com/input-output-hk/cardano-ledger/blob/68a535603c80b7cf53425fd34558e23f9983dfe8/eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Wallet.hs#L506) accepts an _era-specific_ protocol parameters object and, unlike like the Cardano API, does _not_ require that we convert to an era-agnostic protocol parameters object. Since values of the `MinimumUTxO` type already contain era-specific protocol parameters, this means we can avoid the extra complication of record conversion whenever we call:
    - `isBelowMinimumCoinForUTxO`
    - `computeMinimumCoinForUTxO`
    
## Performance

Mainnet (Alonzo):

```sh
$ time curl -X POST http://localhost:8091/v2/wallets/042094088ed439a7b14e811e0781a00185b921c2/payment-fees -d '{"payments":[{"address":"addr1qylgm2dh3vpv07cjfcyuu6vhaqhf8998qcx6s8ucpkly6f8l0dw5r75vk42mv3ykq8vyjeaanvpytg79xqzymqy5acmqgyxuyr","amount":{"quantity":100000000,"unit":"lovelace"}}]}' -H "Content-Type: application/json"
{"deposit":{"quantity":0,"unit":"lovelace"},"estimated_max":{"quantity":187633,"unit":"lovelace"},"estimated_min":{"quantity":169021,"unit":"lovelace"},"minimum_coins":[{"quantity":999978,"unit":"lovelace"}]}
real	0m0,029s
user	0m0,006s
sys	0m0,000s
```

Preprod (Alonzo):
```sh
$ time curl -X POST http://localhost:8090/v2/wallets/1f82e83772b7579fc0854bd13db6a9cce21ccd95/payment-fees \
> -d '{"payments":[{"address":"addr_test1qrtez7vn0d8xp495ggypmu2kyt7tt6qyva2spm0f5a3ewn0v474mcs4q8e9g55yknx3729kyg5dl69x5596ee9tvnynq7ffety","amount":{"quantity":1000000,"unit":"lovelace"}}]}' \
> -H "Content-Type: application/json"
{"deposit":{"quantity":0,"unit":"lovelace"},"estimated_max":{"quantity":169021,"unit":"lovelace"},"estimated_min":{"quantity":169021,"unit":"lovelace"},"minimum_coins":[{"quantity":999978,"unit":"lovelace"}]}
real	0m0,029s
user	0m0,006s
sys	0m0,000s
```

Preview (Babbage):
```sh
$ time curl -X POST http://localhost:8090/v2/wallets/1f82e83772b7579fc0854bd13db6a9cce21ccd95/payment-fees -d '{"payments":[{"address":"addr_test1qrtez7vn0d8xp495ggypmu2kyt7tt6qyva2spm0f5a3ewn0v474mcs4q8e9g55yknx3729kyg5dl69x5596ee9tvnynq7ffety","amount":{"quantity":100000000,"unit":"lovelace"}}]}' -H "Content-Type: application/json"
{"deposit":{"quantity":0,"unit":"lovelace"},"estimated_max":{"quantity":187809,"unit":"lovelace"},"estimated_min":{"quantity":169197,"unit":"lovelace"},"minimum_coins":[{"quantity":995610,"unit":"lovelace"}]}
real	0m0,036s
user	0m0,003s
sys	0m0,003s
```

jonathanknowles committed:
Move `unsafeCoinFromCardanoApiCalculateMinimumUTxOResult` to internal module.

jonathanknowles committed:
Add function `computeMinimumCoinForUTxOCardanoApi`.

This function simply encapsulates the Cardano API minimum UTxO function
in a wallet-friendly interface, acceping and returning wallet-friendly
types.

jonathanknowles committed:
Use `computeMinimumCoinForUTxOCardanoApi` to simplify implementation.

jonathanknowles committed:
Make `unsafeCoinFromCardanoApiCalculateMinimumUTxOResult` an inner function.

This function is now only used by `computeMinimumCoinForUTxOCardanoApi`.

We can safely make it an inner function.

jonathanknowles committed:
Add era-specific `TxOut` conversion functions to `Compatibility.Ledger`.

jonathanknowles committed:
Add function `computeMinimumCoinForUTxOCardanoLedger`.

This function simply encapsulates the Cardano Ledger minimum UTxO
function in a wallet-friendly interface, acceping and returning
wallet-friendly types.

jonathanknowles committed:
Use `computeMinimumCoinForUTxOCardanoLedger` in implementation.

This commit switches from the Cardano API implementation of the minimum
UTxO calculation to the Cardano Ledger implementation.

jonathanknowles committed:
Test that `computeMinimumCoinForUTxO{CardanoApi,CardanoLedger}` are equivalent.

We eventually wish to switch from the Cardano API implementation of the
minimum UTXO calculaion to the Cardano Ledger implementation.

But before we do that, we test that the implementations are equivalent.

jonathanknowles committed:
Do not treat the Babbage era specially in the public implementation.

Now that we are using the Cardano Ledger implementation of the minimum
UTxO calculation, we no longer need to treat the Babbage era specially
in our public implementation.

jonathanknowles committed:
Simplify public function `computeMinimumCoinForUTxO`.

jonathanknowles committed:
Simplify public function `isBelowMinimumCoinForUTxO`.

jonathanknowles committed:
Delete redundant code from public `MinimumUTxO` module.

These functions are no longer used.

jonathanknowles committed:
Remove `prop_computeMinimumCoinForUTxO_evaluation`.

Since our minimum UTxO calculation now relies on Cardano Ledger, we no
longer need to convert from era-specific protocol parameters to the
era-agnostic protocol parameters required by the Cardano API.

Therefore, this property test (which checked that conversions would
never cause run-time errors) is no longer necessary.

jonathanknowles committed:
Use underscores to make function names clearer.

The juxtaposition of "UTxO" followed by another capitalised word is
rather hard to read. This commit introduces underscores to aid
readability.

jonathanknowles committed:
Use `computeMinimumCoinForUTxO_CardanoApi` to simplify property definitions.

jonathanknowles committed:
Simplify property names for `computeMinimumCoinForUTxO`.

jonathanknowles committed:
Revise documentation of `computeMinimumCoinForUTxO` functions.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:
```diff
@@ -177,14 +183,21 @@ spec = do
+        Internal.computeMinimumCoinForUTxO_CardanoLedger
+            minimumUTxO txOut
+  where
+    txOut = TxOut address (TokenBundle coin tokenMap)
```
Q: Why not take a `TokenBundle` directly like in other properties? It seems you needed to define `Arbitrary Coin` specially for this property.

jonathanknowles replied:
> Q: Why not take a `TokenBundle` directly like in other properties? It seems you needed to define `Arbitrary Coin` specially for this property.

Good question! I've simplified this in 9f37edc88a179c7a84fb1c803553337f3b928e0f.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -14,201 +13,95 @@ module Cardano.Wallet.Shelley.MinimumUTxO
 -- | Returns 'True' if and only if the given 'TokenBundle' has a 'Coin' value
 --   that is below the minimum acceptable 'Coin' value.
 --
+-- This function should /only/ be used to validate existing 'Coin' values that
+-- do not need to be modified in any way.
+--
+-- Increasing the 'Coin' value of an output can lead to an increase in the
+-- serialized length of that output, which can in turn lead to an increase in
+-- the minimum required 'Coin' value, since the minimum required 'Coin' value
+-- is dependent on an output's serialized length.
+--
+-- Therefore, even if this function indicates that a given value 'Coin' value
+-- 'c' satisfies the minimum UTxO rule, it should not be taken to imply that
+-- all values greater than 'c' will also satisfy the minimum UTxO rule.
+--
+-- If you need to generate a value that can always safely be increased, use
+-- the 'computeMinimumCoinForUTxO' function instead.
+--
```
:+1:

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO/Internal.hs:
```diff
@@ -0,0 +1,120 @@
+                    $ toAlonzoTxOut txOut Nothing
+            Cardano.ShelleyBasedEraBabbage ->
+                evaluateMinLovelaceOutput pp
+                    $ toBabbageTxOut txOut Nothing
```
👍

Anviking replied:
Possibly we could move the `case era of` to a `toLedger :: ShelleyEra era => W.TxOut -> Ledger.TxOut era`? but it doesn't matter now; we'll see how things turn out in the future

jonathanknowles replied:
I agree. It'd be nice to factor out this pattern match as you suggest. It's something we could do in a follow-up PR.

jonathanknowles replied:
I did give this a try, but couldn't get the types to line up in a reasonable amount of time.

The ledger code has so many re-exports that I just gave up trying to figure out which things to import...

jonathanknowles committed:
Adjust signature of `prop_computeMinimumCoinForUTxO_CardanoApi_CardanoLedger`.

It's simpler for this property to accept a `TokenBundle` instead of a
`TokenMap` and `Coin`.

The `Arbitrary` instance for `TokenBundle` uses the `genTxOutTokenBundle`
generator, which already has good coverage of minimum and maximum values
specifically in the context of a transaction output.

This allows us to get rid of the `Arbitrary` instance for `Coin` (which
in hindsight would have been better to implement with the `genTxOutCoin`
generator).

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3456#discussion_r953637703

jonathanknowles committed:
Miscellaneous formatting fixes.

jonathanknowles committed:
Simplify type signatures of `to{Alonzo,Babbage}TxOut`.

We can use the `Standard{Alonzo,Babbage}` synonyms for greater concision.
